src/test/java/warehouse/CartTest.java

```diff
@@ -33,6 +35,30 @@ void getTotalPrice() {
     }
 
     @Test
-    void addToCartById() {
+    void addToCartById_checkIfTheObjectHasBeenAddedToCart() throws InvalidIdException, EmptyWarehouseException {
+        Product product1 = new Product(DeviceType.SMARTPHONE, "Apple", "xs", "ok", 20, 180, 1200, 1250, 1382);
+        warehouse.addProductToStock(product1);
+        cart.addToCartById(new FilterById(1), warehouse);
+        assertEquals(1, cart.getDevices().size());
+    }
+
+    @Test
+    void addToCartById_cartContainsProducts() throws InvalidIdException, EmptyWarehouseException {
+        Product product1 = new Product(DeviceType.SMARTPHONE, "Apple", "xs", "ok", 10, 180, 1200, 1250, 1382);
+        Product product2 = new Product(DeviceType.SMARTPHONE, "Apple1", "xs", "ok", 10, 180, 1200, 1250, 1382);
+        Product product3 = new Product(DeviceType.SMARTPHONE, "Apple2", "xs", "ok", 10, 180, 1200, 1250, 1382);
+        warehouse.addProductToStock(product1);
+        warehouse.addProductToStock(product2);
+        warehouse.addProductToStock(product3);
+        cart.addToCartById(new FilterById(1), warehouse);
+        cart.addToCartById(new FilterById(2), warehouse);
+        cart.addToCartById(new FilterById(3), warehouse);
+        assertTrue(cart.getDevices().contains(product1));
+    }
+
+    @Test
+    void addToCartById_checkInvalidIdExceptionWhenWarehouseIsEmpty() throws InvalidIdException, EmptyWarehouseException {
```

Comment:
come ti dicevo puoi usare questa exception per testare anche il caso in cui l'id che stai cercando di aggiungere col metodo sia già presente nel cart. Puoi anche partire dal test per poi implementare il relativo controllo nel metodo che dovrebbe appunto lanciare questa eccezione .